strr-web/stores/strr.ts

```diff
@@ -76,6 +78,12 @@ const optionalSin = z
     message: 'Social Insurance Number must be provided in the format 111 111 111'
   })
   .optional()
+const optionalCRABusinessNumber = z
+  .string()
+  .refine(val => val === '' || craBusinessNumberRegex.test(val), {
+    message: 'CRA Business Number must be a 9-digit number'
```

Comment:
I tried to see where this error message coming from but could not find it in ticket or validation spreadsheet. The Business Number to have 9 digits only does not sound right.

Reply:
Check `9-digit business number issued by CRA (BN9)`
https://www2.gov.bc.ca/gov/content/employment-business/business/managing-a-business/permits-licences/businesses-incorporated-companies/business-number

Reply:
BN is 9, but CRA has accounts that are attached to the 9-digit number like RC0001 for GST, RP0001 for Payroll etc.
So your full gst number looks like 123456789RC0001

Reply:
# CRA Business Number

CRA Business Number always starts with a **1** or an **8**.  
Validation rules found here: [Validating Canadian SIN](https://www.straightlineinternational.com/docs/vaildating_canadian_sin.pdf).  
For SIN, but BN follows identical validation except for starting with **1** or **8**.

## Social Insurance Number (SIN)

The Social Insurance Number (**SIN**) was created in 1964 to serve as a client account number in the administration of the Canada Pension Plan (CPP) and Canada's varied employment insurance programs. In 1967, Revenue Canada, now called **Canada Customs and Revenue Agency (CCRA)** or **Canada Revenue Agency (CRA)**, co-opted the SIN and started using it for tax reporting purposes.

_The only legislated uses of the SIN are: Canada Revenue Agency, your employer for Income Tax reporting, banks (with some exceptions), Social Assistance programs, and a few other government and/or tax-related agencies._

## Algorithm

Social Insurance Numbers and Business Numbers are validated via a simple checksum process. This SIN algorithm is commonly known as the **LUHN algorithm** or the **mod-10 algorithm**. It is also used to validate credit card numbers, among other things.

### Example

Let's use this fictitious SIN to demonstrate:

`046 454 286`

| 0 | 4 | 6 | 4 | 5 | 4 | 2 | 8 | 6 |
|---|---|---|---|---|---|---|---|---|
| × | 1 | 2 | 1 | 2 | 1 | 2 | 1 | 2 | 1 |

After multiplying:

`0 8 6 8 5 8 2 7 6`

Notice that 8 × 2 = 16, so you add the digits **1** and **6** together to get **7**. If you get a two-digit number, always sum the digits.

### Final Sum

`0 + 8 + 6 + 8 + 5 + 8 + 2 + 7 + 6 = 50`

If the SIN is valid, this number will be evenly divisible by 10.

`50 / 10 = 5` (Valid SIN)

The SIN algorithm can also be arranged to **generate** as well as **validate**.

## Province First-Digit Usage

The first digit of a SIN indicates the province of registration:

- **1** = NB, NF, NS, PE (Also used for Business Numbers, BN)
- **2** = QC  
- **3** = Not Used? QC?  
- **4** = ON  
- **5** = ON  
- **6** = AB, MB, SK, NT, NU?  
- **7** = BC, YU  
- **8** = Used for Business Numbers (BN)  
- **9** = Immigrants & other temporary SINs  
- **0** = Not Used  

## BN (Business Number)

The **BN** is identical to the SIN but only starts with a **1** or an **8**.

Reply:
After discussion with @rstens , came up with `^\d{9}$|^\d{9}[A-Z]{2}\d{4}$` to validate CRA BN which is 9-digit business number and can be followed by a 2-letter program identifier and a 4-digit reference number.

Reply:
My comment was to point out that it looks like the business requirement is missing from our official ticket and spreadsheet.